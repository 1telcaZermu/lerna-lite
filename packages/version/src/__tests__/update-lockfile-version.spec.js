'use strict';

const path = require('path');
const fs = require('fs-extra');
const yaml = require('js-yaml');

// mocked or stubbed modules
const loadJsonFile = require('load-json-file');

// helpers
const { getPackages } = require('../../../core/src/project');
const initFixture = require('@lerna-test/init-fixture')(__dirname);

const {
  updateClassicLockfileVersion,
  updateTempModernLockfileVersion,
  loadLockfile,
  saveLockfile,
} = require('../lib/update-lockfile-version');

describe('npm classic lock file', () => {
  test('updateLockfileVersion with lockfile v1', async () => {
    const cwd = await initFixture('lockfile-leaf');
    const [pkg] = await getPackages(cwd);

    pkg.version = '2.0.0';

    const returnedLockfilePath = await updateClassicLockfileVersion(pkg);

    expect(returnedLockfilePath).toBe(path.join(pkg.location, 'package-lock.json'));
    expect(Array.from(loadJsonFile.registry.keys())).toStrictEqual(['/packages/package-1']);
    expect(fs.readJSONSync(returnedLockfilePath)).toHaveProperty('version', '2.0.0');
  });

  test('updateClassicLockfileVersion with lockfile v2', async () => {
    const cwd = await initFixture('lockfile-leaf-v2');
    const [pkg] = await getPackages(cwd);

    pkg.version = '2.0.0';

    const returnedLockfilePath = await updateClassicLockfileVersion(pkg);

    expect(returnedLockfilePath).toBe(path.join(pkg.location, 'package-lock.json'));
    expect(Array.from(loadJsonFile.registry.keys())).toStrictEqual(['/packages/package-1']);
    const updatedLockfile = fs.readJSONSync(returnedLockfilePath);
    expect(updatedLockfile).toHaveProperty('version', '2.0.0');
    expect(updatedLockfile).toHaveProperty(['packages', '', 'version'], '2.0.0');
  });

  test('updateClassicLockfileVersion without sibling lockfile', async () => {
    const cwd = await initFixture('lifecycle', false);
    const [pkg] = await getPackages(cwd);

    pkg.version = '1.1.0';

    loadJsonFile.mockImplementationOnce(() => Promise.reject(new Error('file not found')));

    const returnedLockfilePath = await updateClassicLockfileVersion(pkg);

    expect(returnedLockfilePath).toBeUndefined();
    expect(fs.pathExistsSync(path.join(pkg.location, 'package-lock.json'))).toBe(false);
  });
});

describe('npm modern lock file', () => {
  test('call updateNpmLockFileVersion2 for npm lock file in project root', async () => {
    const mockVersion = '2.4.0';
    const cwd = await initFixture('lockfile-version2');
    const rootLockFilePath = path.join(cwd, 'package-lock.json');
    const packages = await getPackages(cwd);

    const lockFileOutput = await loadLockfile(cwd);
    if (lockFileOutput.json) {
      for (const pkg of packages) {
        pkg.version = mockVersion;
        await updateTempModernLockfileVersion(pkg, lockFileOutput);
      }
      await saveLockfile(lockFileOutput);
    }

    expect(lockFileOutput.packageManager).toBe('npm');
    expect(Array.from(loadJsonFile.registry.keys())).toStrictEqual([
      '/packages/package-1',
      '/packages/package-2',
      '/',
    ]);
    expect(fs.readJSONSync(rootLockFilePath)).toMatchSnapshot();
  });
});

describe('pnpm lock file', () => {
  test('call updatePnpmLockFile for pnpm lock file in project root', async () => {
    const mockVersion = '2.5.0';
    const cwd = await initFixture('lockfile-pnpm');
    const packages = await getPackages(cwd);

    // loading lock file should work with/without providing npm client type (2nd arg)
    let lockFileOutput = await loadLockfile(cwd);
    expect(lockFileOutput).not.toBeUndefined();
    expect(lockFileOutput.packageManager).toBe('pnpm');

    lockFileOutput = await loadLockfile(cwd, 'pnpm');
    expect(lockFileOutput).not.toBeUndefined();
    expect(lockFileOutput.packageManager).toBe('pnpm');

    if (lockFileOutput.json) {
      for (const pkg of packages) {
        pkg.version = mockVersion;
        await updateTempModernLockfileVersion(pkg, lockFileOutput);
      }
      await saveLockfile(lockFileOutput);
    }

    expect(Array.from(loadJsonFile.registry.keys())).toStrictEqual([
      '/packages/package-1',
      '/packages/package-2',
      '/packages/package-3',
      '/packages/package-4',
      '/',
    ]);

    expect(`${fs.readFileSync(lockFileOutput.path)}`).toMatchSnapshot();
  });
});
