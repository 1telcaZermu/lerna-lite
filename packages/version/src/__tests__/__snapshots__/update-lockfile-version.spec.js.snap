// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npm modern lock file call updateNpmLockFileVersion2 for npm lock file in project root 1`] = `
Object {
  "dependencies": Object {
    "@my-workspace/package-1": Object {
      "requires": Object {
        "tiny-tarball": "^1.0.0",
      },
      "version": "file:packages/package-1",
    },
    "@my-workspace/package-2": Object {
      "requires": Object {
        "@my-workspace/package-1": "^2.4.0",
      },
      "version": "file:packages/package-2",
    },
  },
  "lockfileVersion": 2,
  "name": "my-workspace",
  "packages": Object {
    "": Object {
      "license": "MIT",
      "name": "my-workspace",
      "workspaces": Array [
        "./packages/package-1",
        "./packages/package-2",
      ],
    },
    "node_modules/package-1": Object {
      "link": true,
      "resolved": "packages/package-1",
    },
    "node_modules/package-2": Object {
      "link": true,
      "resolved": "packages/package-2",
    },
    "packages/package-1": Object {
      "license": "MIT",
      "name": "@my-workspace/package-1",
      "tiny-tarball": Object {
        "integrity": "sha1-u/EC1a5zr+LFUyleD7AiMCFvZbE=",
        "resolved": "https://registry.npmjs.org/tiny-tarball/-/tiny-tarball-1.0.0.tgz",
        "version": "1.0.0",
      },
      "version": "2.4.0",
    },
    "packages/package-2": Object {
      "dependencies": Object {
        "@my-workspace/package-1": "^2.4.0",
      },
      "license": "MIT",
      "name": "@my-workspace/package-2",
      "version": "2.4.0",
    },
  },
  "requires": true,
}
`;

exports[`pnpm lock file call updatePnpmLockFile for pnpm lock file in project root 1`] = `
"lockfileVersion: 5.4
importers:
  .:
    specifiers: {}
  packages/package-1:
    specifiers:
      tiny-tarball: ^1.0.0
    dependencies:
      tiny-tarball: 1.0.0
  packages/package-2:
    specifiers:
      '@my-workspace/package-1': workspace:^2.5.0
    dependencies:
      '@my-workspace/package-1': link:../package-1
packages:
  /tiny-tarball/1.0.0:
    resolution:
      integrity: sha1-u/EC1a5zr+LFUyleD7AiMCFvZbE=
    dev: false
"
`;
